platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,881 @@
+| 7 | Pre-Genesis: Separate app with genesis roster creates key material in state before network officially starts. | <span style="color:green">YES</span> | <span style="color:green">YES</span>                 | <span style="color:green">YES</span> | <span style="color:green">YES</span> | <span style="color:red">NO</span>    | <span style="color:red">NO</span>    | <span style="color:red">NO</span>    | <span style="color:red">HIGH</span>       | Applies to new networks only.  Use Option 1 or 2 for existing networks.            |
+| 8 | Instead of separate app, detect genesis, key the state, network restart from round 0 with keyed state.        | <span style="color:green">YES</span> | <span style="color:green">YES</span>                 | <span style="color:green">YES</span> | <span style="color:green">YES</span> | <span style="color:red">NO</span>    | <span style="color:red">NO</span>    | <span style="color:green">YES</span> | <span style="color:red">HIGH</span>       | Applies to new networks only.  Use Option 1 or 2 for existing networks.            | 
+
+Option 2 was chosen for existing networks and Option 8 was selected for new networks. This decision was made to
+ensure that the first block in the block stream is can be signed by the ledger.
+
+---
+
+## Changes
+
+The changes are presented in the following order:
+
+1. Public API
+2. Core Behaviors
+3. Component Architecture
+4. Configuration
+5. Metrics
+6. Performance
+
+### Public API
+
+The public API changes consist of the following:
+
+* System Transactions
+* State Data Structures
+
+#### System Transactions
+
+The following are new system transactions needed to key rosters and generate ledger signatures.
+
+##### TssMessage System Transaction
+
+```protobuf
+/**
+ * The TssMessageTransaction is used to send a TssMessage to the network for a candidate roster.      
+ */
+message TssMessageTransaction {
+
+  /**
+   * The node id is the id of the node that generated the TssMessage.
+   */
+  uint64 node_id = 1;
+
+  /**
+   * The source roster hash is the hash of the roster containing the node generating the TssMessage.
+   */
+  bytes source_roster_hash = 2;
+
+  /**
+   * The target roster hash is the hash of the roster that the TssMessage is for.
+   */
+  bytes target_roster_hash = 3;
+
+  /**
+   * The share id establishes a global ordering of shares across all shares in the network.  It corresponds to the 
+   * index of the public share in the list returned from the TSS library when the share was created.
+   */
+  uint64 share_id = 4;
+
+
+
+  /**
+   * The tss_message is the raw bytes of the TssMessage generated by the node for the share_id. 
+   */
+  bytes tss_message = 5;
```

Comment:
It's still duplicate data, though, since the `shareId` is included in the TSS message, and you can determine from that share ID which public key is the correct one to use.

> the consensus node's concept of ShareId comes from the order of public keys in the list

I'm personally not convinced that this is the right design decision. It's *possible* to define things in terms of share index instead of share ID, but I'm not sure what exactly we gain from doing so? It seems like an additional layer of terminology on top of an already serviceable unique identifier.

> It's really happenstance if what I am calling the ShareId lines up with any kind of concept of ShareId inside the TSS-Library

If we do go this route, I don't think we should overlap terminology. If it isn't the same share ID, maybe it should be called a `shareIndex` (and also be stored in a smaller `int`)

Reply:
For context:

@edward-swirldslabs , @mxtartaglia-sl , and I discussed this in a call. Key takeaways:

- the duplicate data isn't really a big concern, since it's very little (i.e., we shouldn't let *that alone* have a large impact on the decision)
- there is a goal of not including any protobufs in the Tss Library
- If the concept of a `shareId` is different between the tss library and the consensus node, then there should be separate terminology (with `shareIndex` being a potential option)
- Ed is looking into the potential option of not needing a concept of a share ID on the consensus node *at all*. If this is possible, it makes the whole discussion moot